### PR TITLE
chore: bitwise not for brillig

### PR DESCRIPTION
**Pull request by guipublic**
# Description

Implement bitwise not for brillig

## Problem\*

Related to #1763 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/brillig/brillig_ir.rs [hunk 1]
```diff
@@ -419,17 +419,22 @@ impl BrilligContext {
...
+        let u_max = FieldElement::from(2_i128).pow(&FieldElement::from(bit_size as i128))
+            - FieldElement::one();
         let opcode = BrilligOpcode::BinaryIntOp {
             destination: result,
             op: BinaryIntOp::Sub,
-            bit_size: 1,
-            lhs: one,
-            rhs: condition,
+            bit_size,
+            lhs: self.make_constant(Value::from(u_max)),
```
Comment (ludamad) [inline crates/noirc_evaluator/src/brillig/brillig_ir.rs:436]: We can call deallocate_register on the result of make_constant once we don't need it anymore

File: crates/noirc_evaluator/src/brillig/brillig_ir/debug_show.rs [hunk 2]
```diff
@@ -171,8 +171,8 @@ pub(crate) fn const_instruction(result: RegisterIndex, constant: Value) {
 }
 
 /// Processes a not instruction. Append with "_" as this is a high-level instruction.
-pub(crate) fn not_instruction(condition: RegisterIndex, result: RegisterIndex) {
-    debug_println!("  _NOT {} = !{}", result, condition);
+pub(crate) fn not_instruction(condition: RegisterIndex, bit_size: u32, result: RegisterIndex) {
+    debug_println!("  _NOT({}bits) {} = !{}", bit_size, result, condition);
```
Comment (ludamad) [inline crates/noirc_evaluator/src/brillig/brillig_ir/debug_show.rs:175]: Consistent would be e.g. i32::NOT as that is what I did with other operations

File: crates/noirc_evaluator/src/brillig/brillig_gen/brillig_block.rs [hunk 3]
```diff
@@ -175,18 +175,21 @@ impl<'block> BrilligBlock<'block> {
...
-                );
                 let condition = self.convert_ssa_value(*value, dfg);
                 let result_ids = dfg.instruction_results(instruction_id);
                 let result_register = self
                     .function_context
                     .get_or_create_register(self.brillig_context, result_ids[0]);
-
-                self.brillig_context.not_instruction(condition, result_register);
+                let bit_size = match dfg.type_of_value(*value) {
+                    Type::Numeric(num_type) => match num_type {
```
Comment (ludamad) [inline crates/noirc_evaluator/src/brillig/brillig_gen/brillig_block.rs:184]: Wonder if a bit size helper makes sense

**Review by ludamad — APPROVED** [review]

File: crates/noirc_evaluator/src/brillig/brillig_ir.rs (hunk 1 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/brillig/brillig_ir.rs]: Done

File: crates/noirc_evaluator/src/brillig/brillig_gen/brillig_block.rs (hunk 3 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/brillig/brillig_gen/brillig_block.rs]: I added one

**Review by ludamad — APPROVED** [review]

**Pull request merged**